From f1488d050826e1d72866a63314d2dcd3617ff4ba Mon Sep 17 00:00:00 2001
From: "Nicolas G. Querol" <nicolas.gquerol@gmail.com>
Date: Sat, 27 Jun 2020 18:04:38 +0200
Subject: [PATCH] Add `ns-system-appearance-change-functions' hook

This implements a new hook, effective only on macOS >= 10.14 (Mojave),
that is called when the system changes its appearance (e.g. from light
to dark). Users can then implement functions that take this change
into account, for instance to load a particular theme.

Minor changes are also made to select the right "dark" appearance
(NSAppearanceNameDarkAqua) on macOS versions >= 10.14, the previous one
(NSAppearanceNameVibrantDark) being deprecated.

* src/frame.h (enum ns_appearance_type): Add new
"ns_appearance_dark_aqua" case.

* src/nsfns.m (defun x-create-frame): Use "dark aqua" appearance on
macOS >= 10.14.

* src/nsterm.m:
  - (ns_set_appearance): Use "dark aqua" appearance on
     macOS >= 10.14, reset appearance to the system one
     if `ns-appearance' frame parameter is not set to
     either `dark' or `light'.
  - (initFrameFromEmacs): Use "dark aqua" appearance on
     macOS >= 10.14.
  - (EmacsApp) Add the `systemDidChangeAppearance' private method,
    as well as the appropriate Key-Value Observing calls to update
    the frame's appearance when the system (and thus the app's)
    appearance changes.
  - Add `ns-system-appearance-change-functions' hook variable and
    symbol, to allow users to add functions that react to the
    change of the system's appearance.

Here is an example on how to use this new feature:

    (add-hook 'ns-system-appearance-change-functions
        #'(lambda (appearance)
            (mapc #'disable-theme custom-enabled-themes)
            (pcase appearance
               ('light (load-theme 'tango t))
               ('dark (load-theme 'tango-dark t)))))

The hook is executed once at Emacs startup, and then every time the
system appearance changes.
---
 src/frame.h  |   1 +
 src/nsfns.m  |  13 +++++-
 src/nsterm.m | 122 +++++++++++++++++++++++++++++++++++++++++++++++----
 3 files changed, 126 insertions(+), 10 deletions(-)

diff --git a/src/frame.h b/src/frame.h
index a54b8623e5..46a7c34cb7 100644
--- a/src/frame.h
+++ b/src/frame.h
@@ -70,6 +70,7 @@ #define EMACS_FRAME_H
 enum ns_appearance_type
   {
    ns_appearance_aqua,
+   ns_appearance_dark_aqua,
    ns_appearance_vibrant_dark
   };
 #endif
diff --git a/src/nsfns.m b/src/nsfns.m
index 0f879fe390..5a4dd3a157 100644
--- a/src/nsfns.m
+++ b/src/nsfns.m
@@ -1269,10 +1269,19 @@ Turn the input menu (an NSMenu) into a lisp list for tracking on lisp side.
   store_frame_param (f, Qundecorated, FRAME_UNDECORATED (f) ? Qt : Qnil);
 
 #ifdef NS_IMPL_COCOA
+#ifndef NSAppKitVersionNumber10_14
+#define NSAppKitVersionNumber10_14 1671
+#endif
   tem = gui_display_get_arg (dpyinfo, parms, Qns_appearance, NULL, NULL,
                              RES_TYPE_SYMBOL);
-  FRAME_NS_APPEARANCE (f) = EQ (tem, Qdark)
-    ? ns_appearance_vibrant_dark : ns_appearance_aqua;
+
+  if (NSAppKitVersionNumber >= NSAppKitVersionNumber10_14)
+    FRAME_NS_APPEARANCE(f) =
+      EQ(tem, Qdark) ? ns_appearance_dark_aqua : ns_appearance_aqua;
+  else
+    FRAME_NS_APPEARANCE(f) =
+      EQ(tem, Qdark) ? ns_appearance_vibrant_dark : ns_appearance_aqua;
+
   store_frame_param (f, Qns_appearance, tem);
 
   tem = gui_display_get_arg (dpyinfo, parms, Qns_transparent_titlebar,
diff --git a/src/nsterm.m b/src/nsterm.m
index ac467840a2..0065bf66ab 100644
--- a/src/nsterm.m
+++ b/src/nsterm.m
@@ -2027,16 +2027,35 @@ so some key presses (TAB) are swallowed by the system.  */
 
   if (EQ (new_value, Qdark))
     {
-      window.appearance = [NSAppearance
-                            appearanceNamed: NSAppearanceNameVibrantDark];
-      FRAME_NS_APPEARANCE (f) = ns_appearance_vibrant_dark;
+#if MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
+#ifndef NSAppKitVersionNumber10_14
+#define NSAppKitVersionNumber10_14 1671
+#endif
+      if (NSAppKitVersionNumber >= NSAppKitVersionNumber10_14)
+        {
+          window.appearance = [NSAppearance
+                                  appearanceNamed: NSAppearanceNameDarkAqua];
+          FRAME_NS_APPEARANCE (f) = ns_appearance_dark_aqua;
+        }
+      else
+#endif /* MAC_OS_X_VERSION_MAX_ALLOWED >= 101400 */
+        {
+          window.appearance = [NSAppearance
+                                  appearanceNamed: NSAppearanceNameVibrantDark];
+          FRAME_NS_APPEARANCE (f) = ns_appearance_vibrant_dark;
+        }
     }
-  else
+  else if (EQ (new_value, Qlight))
     {
       window.appearance = [NSAppearance
                             appearanceNamed: NSAppearanceNameAqua];
       FRAME_NS_APPEARANCE (f) = ns_appearance_aqua;
     }
+  else
+    {
+      // Reset window appearance to track the system appearance.
+      window.appearance = nil;
+    }
 #endif /* MAC_OS_X_VERSION_MAX_ALLOWED >= 101000 */
 }
 
@@ -5566,6 +5585,7 @@ Needs to be here because ns_initialize_display_info () uses AppKit classes.
 
    ========================================================================== */
 
+static const void *kEmacsAppKVOContext = &kEmacsAppKVOContext;
 
 @implementation EmacsApp
 
@@ -5811,6 +5831,13 @@ - (void)applicationDidFinishLaunching: (NSNotification *)notification
 	 object:nil];
 #endif
 
+#if defined (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
+  [self addObserver:self
+         forKeyPath:NSStringFromSelector(@selector(effectiveAppearance))
+            options:NSKeyValueObservingOptionInitial|NSKeyValueObservingOptionNew
+            context:&kEmacsAppKVOContext];
+#endif
+
 #ifdef NS_IMPL_COCOA
   /* Some functions/methods in CoreFoundation/Foundation increase the
      maximum number of open files for the process in their first call.
@@ -5849,6 +5876,50 @@ - (void)antialiasThresholdDidChange:(NSNotification *)notification
 #endif
 }
 
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary *)change
+                       context:(void *)context
+{
+#if defined (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
+  if (context == kEmacsAppKVOContext
+      && object == self
+      && [keyPath isEqualToString:
+                    NSStringFromSelector (@selector(effectiveAppearance))])
+    [self systemAppearanceDidChange:
+               [change objectForKey:NSKeyValueChangeNewKey]];
+  else
+#endif /* (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101400 */
+    [super observeValueForKeyPath:keyPath
+                         ofObject:object
+                           change:change
+                          context:context];
+}
+
+- (void)systemAppearanceDidChange:(NSAppearance *)newAppearance
+{
+#if defined (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
+#ifndef NSAppKitVersionNumber10_14
+#define NSAppKitVersionNumber10_14 1671
+#endif
+
+  if (NSAppKitVersionNumber < NSAppKitVersionNumber10_14)
+    return;
+
+  NSAppearanceName appearance_name =
+      [newAppearance bestMatchFromAppearancesWithNames:@[
+        NSAppearanceNameAqua, NSAppearanceNameDarkAqua
+      ]];
+
+  BOOL is_dark_appearance =
+    [appearance_name isEqualToString:NSAppearanceNameDarkAqua];
+
+  pending_funcalls = Fcons (list3 (Qrun_hook_with_args,
+                                   Qns_system_appearance_change_functions,
+                                   is_dark_appearance ? Qdark : Qlight),
+                            pending_funcalls);
+#endif /* (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101400 */
+}
 
 /* Termination sequences:
     C-x C-c:
@@ -6013,6 +6084,14 @@ - (void)applicationDidResignActive: (NSNotification *)notification
   ns_send_appdefined (-1);
 }
 
+- (void)applicationWillTerminate:(NSNotification *)notification
+{
+  NSTRACE ("[EmacsApp applicationWillTerminate:]");
+
+  [self removeObserver:self
+            forKeyPath:NSStringFromSelector(@selector(effectiveAppearance))
+               context:&kEmacsAppKVOContext];
+}
 
 
 /* ==========================================================================
@@ -7492,12 +7571,27 @@ - (instancetype) initFrameFromEmacs: (struct frame *)f
 #if defined (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101000
 #ifndef NSAppKitVersionNumber10_10
 #define NSAppKitVersionNumber10_10 1343
+#endif
+#if MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
+#ifndef NSAppKitVersionNumber10_14
+#define NSAppKitVersionNumber10_14 1671
 #endif
 
-  if (NSAppKitVersionNumber >= NSAppKitVersionNumber10_10
-      && FRAME_NS_APPEARANCE (f) != ns_appearance_aqua)
-    win.appearance = [NSAppearance
-                          appearanceNamed: NSAppearanceNameVibrantDark];
+  if (NSAppKitVersionNumber < NSAppKitVersionNumber10_14)
+#endif /* MAC_OS_X_VERSION_MAX_ALLOWED >= 101400 */
+    if (NSAppKitVersionNumber >= NSAppKitVersionNumber10_10)
+      {
+        if (FRAME_NS_APPEARANCE(f) != ns_appearance_aqua)
+          {
+            win.appearance =
+                [NSAppearance appearanceNamed:NSAppearanceNameVibrantDark];
+          }
+        else
+          {
+            win.appearance =
+                [NSAppearance appearanceNamed:NSAppearanceNameAqua];
+          }
+      }
 #endif
 
 #if defined (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101000
@@ -9606,6 +9700,18 @@ Nil means use fullscreen the old (< 10.7) way.  The old way works better with
 This variable is ignored on macOS < 10.7 and GNUstep.  Default is t.  */);
   ns_use_mwheel_momentum = YES;
 
+  DEFVAR_LISP ("ns-system-appearance-change-functions",
+               Vns_system_appearance_change_functions,
+     doc: /* List of functions to call when the system appearance changes.
+Each function is called with a single argument, which corresponds to the new
+system appearance (`dark' or `light').
+
+This hook is also run once at startup.
+
+This variable is ignored on macOS < 10.14 and GNUstep.  Default is nil.  */);
+  Vns_system_appearance_change_functions = Qnil;
+  DEFSYM(Qns_system_appearance_change_functions, "ns-system-appearance-change-functions");
+
   /* TODO: Move to common code.  */
   DEFVAR_LISP ("x-toolkit-scroll-bars", Vx_toolkit_scroll_bars,
 	       doc: /* SKIP: real doc in xterm.c.  */);
-- 
2.27.0

